nexus/src/app/sagas/snapshot_create.rs

```diff
@@ -558,52 +654,374 @@ async fn ssc_send_snapshot_request(
+    let params = sagactx.saga_params::<Params>()?;
+
+    let pantry_address = sagactx.lookup::<SocketAddrV6>("pantry_address")?;
+    let snapshot_id = sagactx.lookup::<Uuid>("snapshot_id")?;
+
+    let endpoint = format!("http://{}", pantry_address);
+
+    info!(
+        log,
+        "sending snapshot request with id {} for disk {} to pantry endpoint {}",
+        snapshot_id,
+        params.disk_id,
+        endpoint,
+    );
+
+    let client = crucible_pantry_client::Client::new(&endpoint);
+
+    client
+        .snapshot(
+            &params.disk_id.to_string(),
+            &crucible_pantry_client::types::SnapshotRequest {
+                snapshot_id: snapshot_id.to_string(),
+            },
+        )
+        .await
+        .map_err(|e| {
+            ActionError::action_failed(Error::internal_error(&e.to_string()))
+        })?;
+
+    Ok(())
+}
+
+async fn ssc_call_pantry_detach_for_disk(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let log = sagactx.user_data().log();
+    let params = sagactx.saga_params::<Params>()?;
+
+    let pantry_address = sagactx.lookup::<SocketAddrV6>("pantry_address")?;
+
+    let endpoint = format!("http://{}", pantry_address);
+
+    info!(
+        log,
+        "sending detach for disk {} to endpoint {}", params.disk_id, endpoint,
+    );
+
+    let client = crucible_pantry_client::Client::new(&endpoint);
+
+    retry_until_known_result!(log, {
+        client.detach(&params.disk_id.to_string())
+    });
+
+    Ok(())
+}
+
+async fn ssc_detach_disk_from_pantry(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    let (.., authz_disk, db_disk) =
+        LookupPath::new(&opctx, &osagactx.datastore())
+            .disk_id(params.disk_id)
+            .fetch_for(authz::Action::Modify)
+            .await
+            .map_err(ActionError::action_failed)?;
+
+    match db_disk.runtime().attach_instance_id {
+        Some(attach_instance_id) => {
+            if attach_instance_id == *PANTRY_SENTINEL_ID {
+                info!(log, "detaching disk {} from the pantry", params.disk_id);
+
+                osagactx
+                    .datastore()
+                    .disk_update_runtime(
+                        &opctx,
+                        &authz_disk,
+                        &db_disk.runtime().detach(),
+                    )
+                    .await
+                    .map_err(ActionError::action_failed)?;
+            } else {
+                info!(
+                    log,
+                    "disk {} not attached to pantry, it's attached to {}!",
+                    params.disk_id,
+                    attach_instance_id,
+                );
+
+                // XXX what to do here?
```

Comment:
Similarly, I don't know what to do here, and I'm not sure if this could ever really be reached.

At this point, the saga node that "attaches" this disk to the Pantry has succeeded, and all Pantry operations have succeeded. Again, to enter this branch, something must have come along, taken an already attached disk, and attached it somewhere else.